thirdparty/kyaml/runfn/runfn.go

```diff
@@ -192,14 +193,50 @@ func (r RunFns) runFunctions(
 		})
 	}
 
-	var err error
+	inputResources, err := input.Read()
+	if err != nil {
+		return err
+	}
+
+	selectedInput := inputResources
+
+	if !r.Selector.IsEmpty() {
+		err = fnruntime.SetResourceIds(inputResources)
+		if err != nil {
+			return err
+		}
+
+		// select the resources on which function should be applied
+		selectedInput, err = fnruntime.SelectInput(inputResources, []kptfile.Selector{r.Selector}, &fnruntime.SelectionContext{RootPackagePath: r.uniquePath})
+		if err != nil {
+			return err
+		}
+	}
+
+	pb := &kio.PackageBuffer{}
 	pipeline := kio.Pipeline{
-		Inputs:                []kio.Reader{input},
+		Inputs:                []kio.Reader{&kio.PackageBuffer{Nodes: selectedInput}},
 		Filters:               fltrs,
-		Outputs:               outputs,
+		Outputs:               []kio.Writer{pb},
 		ContinueOnEmptyResult: r.ContinueOnEmptyResult,
 	}
 	err = pipeline.Execute()
+	outputResources := pb.Nodes
+
+	if !r.Selector.IsEmpty() {
+		outputResources = fnruntime.MergeWithInput(pb.Nodes, selectedInput, inputResources)
+		deleteAnnoErr := fnruntime.DeleteResourceIds(outputResources)
+		if deleteAnnoErr != nil {
+			return deleteAnnoErr
+		}
+	}
```

Comment:
Now that we have both `render` and `eval` implementation in place, it will be good to analyze if `selector` aware function execution can be shared between `eval` and `render`. Not sure if it can be shared, but if not, having an understanding will help.

Reply:
I made most of the logic shared and moved it to [utils](https://github.com/GoogleContainerTools/kpt/pull/2447/files#diff-614beac5384ae24bb857a6505c25bb2cad2f88cc36126fc8d4e1975e042ae862R58). The rest of the code is specific to eval and render.

Reply:
Ack.